core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -181,26 +178,15 @@ public static Options defaultOptions() {
   private final int mMaxCapacity;
   private final int mMinCapacity;
 
-  // Tracks the resources that are available ordered by lastAccessTime (the first one is
+  // Tracks the resources that are available ordered by lastAccessTime (the head is
   // the most recently used resource).
   @GuardedBy("mLock")
-  private TreeSet<ResourceInternal<T>> mResourceAvailable =
-      new TreeSet<>(new Comparator<ResourceInternal<T>>() {
-        @Override
-        public int compare(ResourceInternal<T> c1, ResourceInternal<T> c2) {
-          if (c1 == c2) {
-            return 0;
-          }
-          if (c1.mLastAccessTimeMs == c2.mLastAccessTimeMs) {
-            return c1.mIdentity - c2.mIdentity;
-          }
-          return (int) (c2.mLastAccessTimeMs - c1.mLastAccessTimeMs);
-        }
-      });
+  private Deque<ResourceInternal<T>> mResourceAvailable;
 
   // Tracks all the resources that are not closed.
-  @GuardedBy("mLock")
-  private HashMap<T, ResourceInternal<T>> mResources = new HashMap<>(32);
+  // The size of the map is guarded by mLock. The mLock should be acquired if you add/delete/iterate
```

Comment:
Good point. 
1. Without the lock to guard (delete/add), i cannot guarantee the hash map size to be within a range (min, max). ConcurrentHashMap doesn't provide functions like UpdateIfCondition(elem, condition).
2. With ConcurrentHashMap, i can retrieve a ResourceInternal corresponding to a Resource without locking because those are readonly operations. No one can remove that element from the map without releasing it.